audio/engine.asm

```diff
@@ -502,115 +772,2383 @@ FadeMusic:
+	call LoadChannel
+	ld hl, CHANNEL_FLAGS1
+	add hl, bc
+	set SOUND_SFX, [hl]
+	call StartChannel
+	pop af
+	dec a
+	jr nz, .start_channels
+	ret
+
+	ld a, [wce5f]
+	bit 5, a	; Stereo flag?
+	jr z, _PlaySFX
+
+	ld hl, wMusicID
+	ld [hl], e
+	inc hl
+	ld [hl], d
+
+	ld hl, SFXPointers
+	add hl, de
+	add hl, de
+	add hl, de
+
+	ld a, [hli]
+	ld [wMusicBank], a
+	ld e, [hl]
+	inc hl
+	ld d, [hl]
+
+	call LoadMusicByte
+	rlca
+	rlca
+	maskbits NUM_MUSIC_CHANS
+	inc a
+
+.loop
+	push af
+	call LoadChannel
+	ld hl, CHANNEL_FLAGS1
+
+	add hl, bc
+	set SOUND_SFX, [hl]
+	push de
+
+	ld a, [wCurChannel]
+	maskbits NUM_MUSIC_CHANS
+	ld e, a
+	ld d, 0
+	ld hl, LRTracks
+	add hl, de
+	ld a, [hl]
+	ld hl, wStereoPanningMask
+	and [hl]
+
+	ld hl, CHANNEL_TRACKS
+	add hl, bc
+	ld [hl], a
+
+	ld hl, CHANNEL_FIELD30
+	add hl, bc
+	ld [hl], a
+	ld a, [wCryTracks]
+
+	cp 2
+	jr c, .skip
+
+	ld a, [wSFXDuration]
+
+	ld hl, CHANNEL_FIELD2E
+	add hl, bc
+	ld [hl], a
+
+	ld hl, CHANNEL_FIELD2F
+	add hl, bc
+	ld [hl], a
+
+	ld hl, CHANNEL_FLAGS2
+	add hl, bc
+	set SOUND_UNKN_0F, [hl]
+
+.skip
+	pop de
+
+	ld hl, CHANNEL_FLAGS1
+	add hl, bc
+	set SOUND_CHANNEL_ON, [hl]
+
+	pop af
+	dec a
+	jr nz, .loop
+
+	ret
+
+	ld hl, wMusicID
+	ld [hl], e
+	inc hl
+	ld [hl], d
+	ld hl, SFXPointers
+	add hl, de
+	add hl, de
+	add hl, de
+	ld a, [hli]
+	ld [wMusicBank], a
+	ld e, [hl]
+	inc hl
+	ld d, [hl]
+	call LoadMusicByte
+	rlca
+	rlca
+	maskbits NUM_MUSIC_CHANS
+	inc a
+
+.cry_channels
+	push af
+	call LoadChannel
+	ld hl, CHANNEL_FLAGS1
+	add hl, bc
+	set SOUND_SFX, [hl]
+	ld hl, CHANNEL_FLAGS2
+	add hl, bc
+	set SOUND_CRY_PITCH, [hl]
+	ld hl, CHANNEL_CRY_PITCH
+	add hl, bc
+	ld a, [wCryPitch]
+	ld [hli], a
+	ld a, [wCryPitch + 1]
+	ld [hl], a
+	ld a, [wCurChannel]
+	maskbits NUM_MUSIC_CHANS
+	cp 3
+	jr nc, .cry_ok
+
+	ld hl, CHANNEL_TEMPO
+	add hl, bc
+	ld a, [wCryLength]
+	ld [hli], a
+	ld a, [wCryLength + 1]
+	ld [hl], a
+
+.cry_ok
+	call StartChannel
+	pop af
 	dec a
+	jr nz, .cry_channels
+
+	ret
+
+
+LoadChannel::
+	call LoadMusicByte
+	inc de
+	and 7
+	ld [wCurChannel], a
+	ld c, a
+	ld b, 0
+	ld hl, ChannelPointers
+	add hl, bc
+	add hl, bc
+	ld c, [hl]
+	inc hl
+	ld b, [hl]
+	ld hl, CHANNEL_FLAGS1
+	add hl, bc
+	res SOUND_CHANNEL_ON, [hl]
+	call ChannelInit
+	ld hl, CHANNEL_MUSIC_ADDRESS
+	add hl, bc
+	call LoadMusicByte
+	ld [hli], a
+	inc de
+	call LoadMusicByte
+	ld [hl], a
+	inc de
+	ld hl, CHANNEL_MUSIC_ID
+	add hl, bc
+	ld a, [wMusicID]
+	ld [hli], a
+	ld a, [wMusicID + 1]
+	ld [hl], a
+	ld hl, CHANNEL_MUSIC_BANK
+	add hl, bc
+	ld a, [wMusicBank]
+	ld [hl], a
+	ret
+
+
+ChannelInit:
+	push de
+	xor a
+	ld hl, CHANNEL_MUSIC_ID
+	add hl, bc
+	ld e, CHANNEL_STRUCT_LENGTH
+
+.loop
+	ld [hli], a
+	dec e
 	jr nz, .loop
+
+	ld hl, CHANNEL_TEMPO
+	add hl, bc
 	xor a
-	; TODO
+	ld [hli], a
+	inc a
+	ld [hl], a
+	ld hl, CHANNEL_NOTE_LENGTH
+	add hl, bc
+	ld [hl], a
+	pop de
+	ret
+
+
+LoadMusicByte::
+	ld a, [wMusicBank]
+	call _LoadMusicByte
+	ret
+
+FrequencyTable:
```

Comment:
The various data below here could be extracted into individual `audio/` files; check the ones that exist in pokegold.